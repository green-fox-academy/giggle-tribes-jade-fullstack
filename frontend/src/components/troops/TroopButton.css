div.troopsButton {
  justify-content: center;
  display: flex;
  align-items: center;
}

button.MuiButtonBase-root {
  justify-self: center;
  display: flex;
  justify-content: space-around;
  align-items: center;
  background-color: #009587;
  color: white;
  width: 20em;

  border-radius: 1em;
  z-index: 2;
}
button.MuiButtonBase-root:hover {
  background-color: #009587;
}

button.MuiButtonBase-root p {
  color: white;
  text-transform: none;
  font-size: 1.5em;
}

button.MuiButtonBase-root.toop.button img {
  height: 4em;
  width: 4em;
}

.costLabel {
  display: flex;
  align-items: center;
  background-color: #ff7743;
  border-bottom-right-radius: 1em;
  border-top-right-radius: 1em;
  width: 9em;
  height: 4em;
  position: relative;
  right: 1.1em;
  z-index: 1;
}

.costLabel p {
  color: white;
  text-align: center;
  padding-left: 1.5em;
}

img.inlineIcon {
  height: 15px;
  max-width: 15px;
}
