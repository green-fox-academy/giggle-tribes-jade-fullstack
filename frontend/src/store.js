import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import { userReducer } from './reducers/userReducer';
import { kingdomReducer } from './reducers/kingdomReducer';
import { buildingsReducer } from './reducers/buildingsReducer';
import { resourcesReducer } from './reducers/resourcesReducer';
import { troopsReducer } from './reducers/troopsReducer';
import { tokenReducer } from './reducers/tokenReducer';
import { errorReducer } from './reducers/errorReducer';

function root(state = {}, action) {
  return {
    user: userReducer(state.user, action),
    kingdom: kingdomReducer(state.kingdom, action),
    buildings: buildingsReducer(state.buildings, action),
    resources: resourcesReducer(state.resources, action),
    troops: troopsReducer(state.troops, action),
    token: tokenReducer(state.token, action),
    error: errorReducer(state.error, action),
  };
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(
  root,
  /* preloadedState, */ composeEnhancers(applyMiddleware(thunk))
);

export { store };
