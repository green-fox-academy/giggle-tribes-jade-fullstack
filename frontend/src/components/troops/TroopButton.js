import React from 'react';
import { Button } from '@material-ui/core';
import PropTypes from 'prop-types';

import './TroopButton.css';
import Troop from '../../assets/icons/troops.svg';

export default function TroopButton({ troopSummary }) {
  return (
    <Button className="troops button" onClick={() => {}}>
      <img className="troops button image" src={Troop} alt="troop icon" />
      <p>{troopSummary.amount + ' Troop level ' + troopSummary.level}</p>
    </Button>
  );
}

TroopButton.propTypes = {
  troopSummary: PropTypes.any.isRequired,
};
