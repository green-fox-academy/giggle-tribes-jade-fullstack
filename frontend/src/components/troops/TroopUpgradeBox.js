import React from 'react';
import { connect } from 'react-redux';
import { Button } from '@material-ui/core';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import PropTypes from 'prop-types';

import './TroopButton.css';
import { upgradeTroopAction } from '../../actions/TroopsActions';
import { setErrorAction } from '../../actions/ErrorActions';

function TroopUpgradeBox({
  open,
  setOpen,
  level,
  troopAmount,
  kingdom,
  upgrade,
  resources,
  buildings,
  setError,
}) {
  const [amount, setAmount] = React.useState(0);

  const handleClose = () => {
    setOpen(false);
  };

  const upgradeTroopClickOn = () => {
    const troopUpgradeConditions = {
      upgradeCost: 10,
      goldAmount: resources.find(resource => resource.type === 'gold').amount,
      academyLevel:
        buildings.length > 0
          ? buildings.find(e => e.type === 'academy').level
          : 1,
    };

    if (
      troopUpgradeConditions.goldAmount >= troopUpgradeConditions.upgradeCost &&
      level < troopUpgradeConditions.academyLevel &&
      amount <= troopAmount
    ) {
      upgrade(kingdom, amount, level);
    } else if (
      troopUpgradeConditions.goldAmount < troopUpgradeConditions.upgradeCost
    ) {
      setError("You don't have enough money.");
    } else if (amount > troopAmount) {
      setError(
        `Amount was too much, you have ${troopAmount} troops in that troop level`
      );
    } else {
      setError('Upgrade is not allowed, academy level too low');
    }
    handleClose();
  };

  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">Troop Upgrade</DialogTitle>
      <DialogContent>
        <DialogContentText>
          Please provide amount of troops to be upgraded and click upgrade.
        </DialogContentText>
        <TextField
          autoFocus
          margin="dense"
          id="name"
          label="Amount"
          type="number"
          InputProps={{ inputProps: { min: 0 } }}
          onChange={e => {
            setAmount(e.target.value);
          }}
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          Cancel
        </Button>
        <Button
          className="TroopUpgradeBTN"
          onClick={upgradeTroopClickOn}
          color="primary"
        >
          Upgrade
        </Button>
      </DialogActions>
    </Dialog>
  );
}

TroopUpgradeBox.propTypes = {
  open: PropTypes.bool.isRequired,
  setOpen: PropTypes.func.isRequired,
  level: PropTypes.number.isRequired,
  troopAmount: PropTypes.number.isRequired,
  kingdom: PropTypes.string.isRequired,
  troops: PropTypes.array.isRequired,
  resources: PropTypes.array.isRequired,
  buildings: PropTypes.array.isRequired,
  setError: PropTypes.func.isRequired,
};

const mapStateToProps = ({ kingdom, troops, resources, buildings }) => {
  return { kingdom, troops, resources, buildings };
};

const mapDispatchToProps = dispatch => {
  return {
    upgrade: (kingdomId, amount) => {
      dispatch(upgradeTroopAction(kingdomId, amount));
    },
    setError: error => {
      dispatch(setErrorAction(error));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TroopUpgradeBox);
