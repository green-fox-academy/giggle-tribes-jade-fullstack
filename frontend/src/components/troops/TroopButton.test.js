import React from 'react';
import { Provider } from 'react-redux';
import { render, unmountComponentAtNode } from 'react-dom';
import { act, Simulate } from 'react-dom/test-utils';

import TroopButton from './TroopButton';
import TroopUpgradeBox from './TroopUpgradeBox';
import { store } from '../../store';

let container;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('CreateTroopButton tests', () => {
  it('dialog must be rendered when button is clicked', async () => {
    const handleClickOpen = jest.fn();
    await act(async () => {
      render(
        <Provider store={store}>
          <TroopButton troopSummary={{ amount: 1, level: 1 }} />
        </Provider>,
        container
      );
    });
    const button = document.querySelector('button');

    Simulate.click(button);
    const isDialog = document.getElementsByClassName('MuiDialog-root')
      ? true
      : false;
    expect(isDialog).toBe(true);
  });

  it('matches snapeshot', async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <TroopButton troopSummary={{ amount: 1, level: 1 }} />
        </Provider>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
  });
});
