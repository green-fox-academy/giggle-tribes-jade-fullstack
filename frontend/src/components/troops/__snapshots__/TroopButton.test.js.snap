// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTroopButton tests matches snapeshot 1`] = `"<div class=\\"troopsButton\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text troops button MuiButton-textSizeSmall MuiButton-sizeSmall\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\"><img class=\\"troops button image\\" src=\\"troops.svg\\" alt=\\"troop icon\\"><p>1 Troop level 1</p></span><span class=\\"MuiTouchRipple-root\\"></span></button><div class=\\"costLabel\\"><p>Upgrade <br>10 <img class=\\"inlineIcon\\" src=\\"GoldIcon.svg\\" alt=\\"attack icon\\"> per troop</p></div><div></div></div>"`;
